### ISSUE TITLE
WImage3D shows white rectangle for some images

### ISSUE THREAD

**Issue by dustsigns** [opencv/opencv#9976]
##### System information (version)
- OpenCV => 3.3
- Operating System / Platform => Ubuntu 16.04 with QT 5.5 and VTK 6.2
- Compiler => g++ 5.4.0

##### Detailed description

When visualizing images with `WImage3D`, some images appear as white rectangles, while (most) others display just fine. I cannot see any noticable difference between the images that would explain this. Consider the following example code:

##### Steps to reproduce

```
#include <iostream>
#include <opencv2/viz.hpp>
#include <opencv2/highgui.hpp>

using namespace std;
using namespace cv;
using namespace viz;

int main(const int argc, const char * const argv[])
{
  if (argc != 2)
  {
    cerr << "Usage: " << argv[0] << " <input image>" << endl;
    return 1;
  }
  const Mat image = imread(argv[1]);
  assert(!image.empty());
  Viz3d visualization("Test");
  WImage3D viz_image(image, Size2d(1, 1));
  visualization.showWidget("Image", viz_image);
  visualization.spin();
}
```

When using Lenna (https://en.wikipedia.org/wiki/File:Lenna.png) as an input image, everything works fine:

![lenna3d](https://user-images.githubusercontent.com/11800082/32212439-0009ebc4-be17-11e7-9bc6-06381e843ade.PNG)

However, some other images, like im2.png from http://vision.middlebury.edu/stereo/data/scenes2003/newdata/cones/cones-png-2.zip look like this:

![white3d](https://user-images.githubusercontent.com/11800082/32212475-251c04a6-be17-11e7-9fe6-7992ff2f36ba.PNG)

`cv::imshow` and other functions work fine on this image. It seems to be a PNG with 3 channels (RGB), just like Lenna.

The following error is output to the console when executing the above code with im2.png, but not with Lenna.png:

> ERROR: In /build/vtk6-YpT4yb/vtk6-6.2.0+dfsg1/Rendering/OpenGL/vtkOpenGLTexture.cxx, line 415
> vtkOpenGLTexture (0xb12390): failed at glTexImage2D 1 OpenGL errors detected
>   0 : (1282) Invalid operation

**Comment by alalek** [comment]
Probably image size if not power of 2 (some OpenGL runtimes doesn't support custom sizes): im2.png size is 450x375. Lena.jpg size is 512x512.
Try to call resize before passing image to WImage3D.

**Comment by dustsigns** [comment]
For a picture from the foreman sequence (352x288), it works just fine. So power of two does not seem to be the (sole) issue.

**Comment by LaurentBerger** [comment]
@dustsigns using your program I haven't got any problem for img2.png using opencv 3.3-dev and vtk 8.1 windows 10 visual 2015.
![screenshot-1509523074](https://user-images.githubusercontent.com/3591626/32265333-db66c770-bee2-11e7-9254-b0c21ed8abf2.jpg)

**Comment by dustsigns** [comment]
Maybe the issue is platform-specific, like, e.g., #9759. Thank you for trying it on a different platform.

I already considered reinstalling my Ubuntu using the latest non-stable version (18.04 pre-alpha currently), but they are still only providing VTK 6.x packages there and I'm not sure whether manually building VTK 8 would make this problem (or #9759) go away to begin with. If anyone is using a different platform, but a newer VTK than me, I would be happy to hear whether or not a newer VTK version fixes this issue or #9759.

**Comment by dustsigns** [comment]
After @alalek 's hint regarding size, I tried to figure out what the limitation in my setup was. It seems to be that each image dimension must be at least a multiple of four. In case anyone else has this problem in the future, this is how I worked around this issue:

```
#include <iostream>
#include <opencv2/viz.hpp>
#include <opencv2/highgui.hpp>
#include <opencv2/imgproc.hpp>

using namespace std;
using namespace cv;
using namespace viz;

int main(const int argc, const char * const argv[])
{
  if (argc != 2)
  {
    cerr << "Usage: " << argv[0] << " <input image>" << endl;
    return 1;
  }
  const Mat image = imread(argv[1]);
  assert(!image.empty());
  Mat fixed_image;
  copyMakeBorder(image, fixed_image, 0, image.rows & 3, 0, image.cols & 3, BORDER_REFLECT_101, Scalar()); //Workaround: Make sure each image dimension is a multiple of four
  Viz3d visualization("Test");
  visualization.showWidget("Image", viz_image);
  visualization.spin();
}

```
